[Cosmos] CosmosDB asynchronous client (#21404)
Azure/azure-sdk-for-python: sdk/cosmos/azure-cosmos/README.md

    container = database.get_container_client(container_name)
    for i in range(1, 10):
        await container.upsert_item({
                'id': 'item{0}'.format(i),
                'productName': 'Widget',
                'productModel': 'Model {0}'.format(i)
            }
        )
```

### Queries with the asynchronous client

Queries work the same way for the most part, and results can be directly iterated on, but because queries made by the asynchronous client return AsyncIterable objects, results can't be cast into lists directly; instead, if you need to create lists from your results, use Python's list comprehension to populate a list:

```Python
from azure.cosmos.aio import CosmosClient
import os

url = os.environ['ACCOUNT_URI']
key = os.environ['ACCOUNT_KEY']
client = CosmosClient(url, credential=key)
database_name = 'testDatabase'
database = client.get_database_client(database_name)
container_name = 'products'
container = database.get_container_client(container_name)

async def create_lists():
    results = await container.query_items(
            query='SELECT * FROM products p WHERE p.productModel = "Model 2"',

>             enable_cross_partition_query=True)

Author: does this mean it should be enabled by default if no partition key value is passed in by the user?

For more information on TTL, see [Time to Live for Azure Cosmos DB data][cosmos_ttl].

### Using the asynchronous client

> The asynchronous cosmos client looks and works in a very similar fashion to the already existing client, with the exception of its package within the sdk and the need of using async/await keywords in order to interact with it.

Reviewer: ```suggestion
The asynchronous cosmos client is a separate client that looks and works in a similar fashion to the existing synchronous client. However, the async client needs to be imported separately and its methods need to be used with the async/await keywords.
```

```Python
from azure.cosmos.aio import CosmosClient
import os

> url = os.environ['ACCOUNT_URI']

Reviewer: Constants such as these should be in all caps: https://www.python.org/dev/peps/pep-0008/#constants

KEY = os.environ['ACCOUNT_KEY']
client = CosmosClient(URL, credential=KEY)
DATABASE_NAME = 'testDatabase'
database = client.get_database_client(DATABASE_NAME)
CONTAINER_NAME = 'products'
container = database.get_container_client(CONTAINER_NAME)

async def create_items():

>     for i in range(1, 10):

Reviewer: ```suggestion
    for i in range(10):
```

        await container.upsert_item({
                'id': 'item{0}'.format(i),
                'productName': 'Widget',
                'productModel': 'Model {0}'.format(i)
            }
        )

>     await client.close()

Reviewer: ```suggestion
    await client.close() # the async client must be closed manually if it's not initialized in a with statement
```

```

> It is also worth pointing out that the asynchronous client has to be closed manually after its use, either by initializing it using async with or calling the close() method directly like shown above.

Reviewer: @simorenoh 
> closed manually after its use
Does it need to be closed after every function it's used in (like above), or can the client be passed around safely when doing things async?

Author: It can be passed around safely, but needs to be manually disposed of once you're done using it - the alternative to this being to use the client using a `with` statement:
`async with CosmosClient(url, key) as client:` and then starting your cosmos db logic within that context

Reviewer: @simorenoh  What happens if these async clients are not closed properly? Memory leak or something else?

Author: I would have to test, but I assume that is the case - you also get an error in your code stating the client was not properly disposed of

```Python
from azure.cosmos.aio import CosmosClient
import os

url = os.environ['ACCOUNT_URI']
key = os.environ['ACCOUNT_KEY']
database_name = 'testDatabase'

> container_name = 'products'

Reviewer: ditto. I was a bit confused when I saw these being referred in the function below without realizing they are constants.

> async with CosmosClient(URL, credential=KEY) as client:

Reviewer: ```suggestion
async with CosmosClient(URL, credential=KEY) as client: # the with statement will automatically close the async client
```

    database = client.get_database_client(DATABASE_NAME)
    container = database.get_container_client(CONTAINER_NAME)

>     for i in range(1, 10):

Reviewer: ```suggestion
    for i in range(10):
```

        await container.upsert_item({
                'id': 'item{0}'.format(i),
                'productName': 'Widget',
                'productModel': 'Model {0}'.format(i)
            }

>         )

Reviewer: Do we need to close the client here too?

await client.close()

Author: this is that second way to initialize I had mentioned above - the way it is done here, it gets disposed once the `with` statement gets exited

```

### Queries with the asynchronous client

> Queries work the same way for the most part, with one exception being the absence of the `enable_cross_partition` flag in the request; queries without a specified partition key value will now by default atempt to do a cross partition query. Results can be directly iterated on, but because queries made by the asynchronous client return AsyncIterable objects, results can't be cast into lists directly; instead, if you need to create lists from your results, use Python's list comprehension to populate a list:

Reviewer: ```suggestion
Queries work the same way for the most part, with one exception being the absence of the `enable_cross_partition` flag in the request; queries without a specified partition key value will now by default attempt to do a cross partition query. Results can be directly iterated on, but because queries made by the asynchronous client return AsyncIterable objects, results can't be cast into lists directly; instead, if you need to create lists from your results, use Python's list comprehension to populate a list:
```

Reviewer: 1. Is this the reason that we don't have an await when querying? Should we make that more explicit?
2. Why do we need the word "async" again in the function
3. IIRC point queries operate differently. Should we add an example and explain the differentiation?

Author: 1. Yeah this is that point I had brought up with you before, I feel like I tried to explain it but I'm not sure it's completely understandable... let's talk on this to reach a conclusion.
2. We need the word async in the function definition not because of the query, but because we do an `async for` loop right after - any function that has either `await` or `async` within has to be defined as async by python
3. This is the case even for the normal sync client, and I believe should be documented in the actual Cosmos docs (not just SDK), found this for instance: https://devblogs.microsoft.com/cosmosdb/point-reads-versus-queries/

> Queries work the same way for the most part, with one exception being the absence of the `enable_cross_partition` flag in the request; queries without a specified partition key value will now by default attempt to do a cross partition query. Results can be directly iterated on, but because queries made by the asynchronous client return AsyncIterable objects, results can't be cast into lists directly; instead, if you need to create lists from your results, use Python's list comprehension to populate a list:

Reviewer: ```suggestion
Unlike the synchronous client, the async client does not have an `enable_cross_partition` flag in the request. Queries without a specified partition key value will attempt to do a cross partition query by default. 

Query results can be iterated, but query results return an asynchronous iterator. This means that each object from the iterator is already being "awaited" and does not contain that actual query result, only an awaited object to asynchronously retrieve it. 

These results can't be cast into lists directly; instead, if you need to create lists from your results, use Python's list comprehension to populate a list:
```

Reviewer: @simorenoh  Not sure if "awaited object" is the correct way to describe it. We should probably also add an example of what the results iterator returns.

```Python
from azure.cosmos.aio import CosmosClient
import os

> url = os.environ['ACCOUNT_URI']

Reviewer: ditto on the constants

key = os.environ['ACCOUNT_KEY']
client = CosmosClient(url, credential=key)
database_name = 'testDatabase'
database = client.get_database_client(database_name)
container_name = 'products'
container = database.get_container_client(container_name)

> async def create_lists():

Reviewer: Do we need to close the client in this example too? I'm not seeing "with" being used either.

    results = container.query_items(
            query='SELECT * FROM products p WHERE p.productModel = "Model 2"')

>     # Iterating directly on results

Reviewer: ```suggestion
    # Iterating directly on results
    item_list = []
```

    async for item in results:

>         print(item)

Reviewer: ```suggestion
        item_list.append(item)
```

>     # Making a list from the results

Reviewer: ```suggestion
    # Asynchronously creates a complete list of the actual query results
```

    item_list = [item async for item in results]

> ```

Reviewer: ```suggestion
await client.close()
```
```

```

It is also worth pointing out that the asynchronous client has to be closed manually after its use, either by initializing it using async with or calling the close() method directly like shown above.

```Python
from azure.cosmos.aio import CosmosClient
import os

URL = os.environ['ACCOUNT_URI']
KEY = os.environ['ACCOUNT_KEY']
DATABASE_NAME = 'testDatabase'
CONTAINER_NAME = 'products'

async with CosmosClient(URL, credential=KEY) as client: # the with statement will automatically close the async client
    database = client.get_database_client(DATABASE_NAME)
    container = database.get_container_client(CONTAINER_NAME)
    for i in range(10):
        await container.upsert_item({
                'id': 'item{0}'.format(i),
                'productName': 'Widget',
                'productModel': 'Model {0}'.format(i)
            }
        )
```

### Queries with the asynchronous client

Unlike the synchronous client, the async client does not have an `enable_cross_partition` flag in the request. Queries without a specified partition key value will attempt to do a cross partition query by default. 

> Query results can be iterated, but query results return an asynchronous iterator. This means that each object from the iterator is already being "awaited" and does not contain that actual query result, only an awaitable object to asynchronously retrieve it. 

Author: ```suggestion
Query results can be iterated, but the query's raw output returns an asynchronous iterator. This means that each object from the iterator is an awaitable object, and does not yet contain the true query result. In order to obtain the query results you can use an async for loop, which awaits each result as you iterate on the object, or manually await each query result as you iterate over the asynchronous iterator.
```

> These results can't be cast into lists directly; instead, if you need to create lists from your results, use an async for loop or Python's list comprehension to populate a list:

Author: ```suggestion
Since the query results are an asynchronous iterator, they can't be cast into lists directly; instead, if you need to create lists from your results, use an async for loop or Python's list comprehension to populate a list:
```
